app/src/main/java/org/kiwix/kiwixmobile/localFileTransfer/WifiDirectManager.kt

```diff
@@ -239,49 +222,72 @@ class WifiDirectManager @Inject constructor(
     })
   }
 
-  private fun performHandshakeWithSelectedPeerDevice() {
+  private fun performHandshakeWith(groupInfo: WifiP2pInfo) {
     if (BuildConfig.DEBUG) {
       Log.d(TAG, "Starting handshake")
     }
-    peerGroupHandshakeAsyncTask = PeerGroupHandshakeAsyncTask(this)
-      .also { it.execute() }
+    lifecycleCoroutineScope.launch {
+      val peerGroupHandshake = if (isFileSender) {
+        SenderHandShake(this@WifiDirectManager, groupInfo)
+      } else {
+        ReceiverHandShake(this@WifiDirectManager, groupInfo)
+      }
+      val inetAddress = peerGroupHandshake.handshake()
+      if (inetAddress != null) {
+        setClientAddress(inetAddress, groupInfo)
+      } else {
+        if (BuildConfig.DEBUG) {
+          Log.d(TAG, "InetAddress is null")
+        }
+        onFileTransferAsyncTaskComplete(false)
+        activity.toast(R.string.connection_refused)
+      }
+    }
   }
 
   val totalFilesForTransfer: Int
     get() = filesForTransfer.size
 
   fun getFilesForTransfer() = filesForTransfer
 
-  fun setFilesForTransfer(fileItems: ArrayList<FileItem>) {
+  fun setFilesForTransfer(fileItems: List<FileItem>) {
     filesForTransfer = fileItems
   }
 
-  val zimStorageRootPath
-    get() = sharedPreferenceUtil.prefStorage + "/Kiwix/"
+  val zimStorageRootPath get() = sharedPreferenceUtil.prefStorage + "/Kiwix/"
 
   fun getFileReceiverDeviceAddress() = fileReceiverDeviceAddress
 
-  fun setClientAddress(clientAddress: InetAddress) {
-
+  private fun setClientAddress(clientAddress: InetAddress, groupInfo: WifiP2pInfo) {
```

Comment:
as I said in another comment, this method can be deleted along with `selectedPeerDeviceInetAddress` and you can just use `startFileTransfer`